src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,395 @@
+                }
+                _numPendingRemoves = 0;
+
+                // Fill in nulls
+                // This is O(1), Goes through each of the 63 possible handles once.
+                for (int i = 0; i < _numUserWaits; i++)
+                {
+                    if (_registeredWaitHandles[i] == null)
+                    {
+                        for (int j = _numUserWaits - 1; j > i; j--)
+                        {
+                            if(_registeredWaitHandles[j] != null)
+                            {
+                                _registeredWaitHandles[i] = _registeredWaitHandles[j];
+                                _registeredWaitHandles[j] = null;
+                                _waitHandles[i + 1] = _waitHandles[j + 1];
+                                _waitHandles[j + 1] = null;
+                                _numUserWaits = j;
+                                break;
+                            }
+                        }
+                        if (_registeredWaitHandles[i] == null)
+                        {
+                            _numUserWaits = i - 1;
+                            break;
+                        }
+                    }
+                }
+
+                // Recalculate Timeout
+                int timeout = Timeout.Infinite;
+                for (int i = 0; i < _numUserWaits; i++)
+                {
+                    if (timeout == Timeout.Infinite)
+                    {
+                        timeout = _registeredWaitHandles[i].Timeout;
+                    }
+                    else
+                    {
+                        timeout = Math.Min(_registeredWaitHandles[i].Timeout, timeout);
+                    }
+                }
+                _currentTimeout = timeout;
+                _registeredHandlesLock.Release();
+                _removesLock.Release();
+            }
+            
+            /// <summary>
+            /// Queue a call to <see cref="CompleteWait(object)"/> on the ThreadPool.
+            /// </summary>
+            /// <param name="registeredHandle">The handle that completed.</param>
+            /// <param name="timedOut">Whether or not the wait timed out.</param>
+            private void QueueWaitCompletion(RegisteredWaitHandle registeredHandle, bool timedOut)
+            {
+                ThreadPool.QueueUserWorkItem(CompleteWait, new CompletedWaitHandle(registeredHandle, timedOut));
+            }
+
+            /// <summary>
+            /// Process the completion of a user-registered wait.
+            /// </summary>
+            /// <param name="state">A <see cref="CompletedWaitHandle"/> object representing the wait completion.</param>
+            private void CompleteWait(object state)
+            {
+                CompletedWaitHandle handle = (CompletedWaitHandle)state;
+                handle.CompletedHandle.CanUnregister.Reset();
+                handle.CompletedHandle.PerformCallback(handle.TimedOut);
```

Comment:
It looks like this is already covered actually, but this may also be a good test to add